providers/vsphere/resources/host.go

```diff
@@ -0,0 +1,451 @@
+	hostInventoryPath string
+}
+
+func (v *mqlVsphereVmnic) esxiClient() (*resourceclient.Esxi, error) {
+	conn := v.MqlRuntime.Connection.(*connection.VsphereConnection)
+	return esxiClient(conn, v.hostInventoryPath)
+}
+
+func (v *mqlVsphereVmnic) details() (map[string]interface{}, error) {
+	name := v.Name.Data
+
+	esxiClient, err := v.esxiClient()
+	if err != nil {
+		return nil, err
+	}
+
+	return esxiClient.ListNicDetails(name)
+}
+
+func (v *mqlVsphereHost) vmknics() ([]interface{}, error) {
+	esxiClient, err := v.esxiClient()
+	if err != nil {
+		return nil, err
+	}
+	vmknics, err := esxiClient.Vmknics()
+	if err != nil {
+		return nil, err
+	}
+
+	mqlVmknics := make([]interface{}, len(vmknics))
+	for i := range vmknics {
+		entry := vmknics[i]
+		mqlVswitch, err := CreateResource(v.MqlRuntime, "vsphere.vmknic", map[string]*llx.RawData{
+			"name":       llx.StringData(entry.Properties["Name"].(string)),
+			"properties": llx.DictData(entry.Properties),
+			"ipv4":       llx.ArrayData(entry.Ipv4, types.Dict),
+			"ipv6":       llx.ArrayData(entry.Ipv6, types.Dict),
+			"tags":       llx.ArrayData(convert.SliceAnyToInterface(entry.Tags), types.String),
+		})
+		if err != nil {
+			return nil, err
+		}
+		mqlVmknics[i] = mqlVswitch
+	}
+
+	return mqlVmknics, nil
+}
+
+func (v *mqlVsphereHost) packages() ([]interface{}, error) {
+	esxiClient, err := v.esxiClient()
+	if err != nil {
+		return nil, err
+	}
+	vibs, err := esxiClient.Vibs()
+	if err != nil {
+		return nil, err
+	}
+
+	mqlPackages := make([]interface{}, len(vibs))
+	for i := range vibs {
+		vib := vibs[i]
+
+		// parse timestamps in format "2020-07-16"
+		format := "2006-01-02"
+		var creationDate *time.Time
+		parsedCreation, err := time.Parse(format, vib.CreationDate)
+		if err != nil {
+			return nil, errors.New("cannot parse vib creationDate: " + vib.CreationDate)
+		}
+		creationDate = &parsedCreation
+
+		var installDate *time.Time
+		parsedInstall, err := time.Parse(format, vib.InstallDate)
+		if err != nil {
+			return nil, errors.New("cannot parse vib installDate: " + vib.InstallDate)
+		}
+		installDate = &parsedInstall
+
+		mqlVib, err := CreateResource(v.MqlRuntime, "esxi.vib", map[string]*llx.RawData{
+			"id":              llx.StringData(vib.ID),
+			"name":            llx.StringData(vib.Name),
+			"acceptanceLevel": llx.StringData(vib.AcceptanceLevel),
+			"creationDate":    llx.TimeDataPtr(creationDate),
+			"installDate":     llx.TimeDataPtr(installDate),
+			"status":          llx.StringData(vib.Status),
+			"vendor":          llx.StringData(vib.Vendor),
+			"version":         llx.StringData(vib.Version),
+		})
+		if err != nil {
+			return nil, err
+		}
+		mqlPackages[i] = mqlVib
+	}
+
+	return mqlPackages, nil
+}
+
+func (v *mqlVsphereHost) acceptanceLevel() (string, error) {
+	esxiClient, err := v.esxiClient()
+	if err != nil {
+		return "", err
+	}
+	return esxiClient.SoftwareAcceptance()
+}
+
+func (v *mqlVsphereHost) kernelModules() ([]interface{}, error) {
```

Comment:
Out of curiosity: Why do we not need a safe type here but in other func signatures?

Reply:
arrays are still following the interface-list signature, which we can tackle in a future version. The change requires a bit more than just returning the safe type